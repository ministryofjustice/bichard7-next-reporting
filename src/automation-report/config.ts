import type { ApiConfig, S3Config } from "../shared/types"

export type AutomationReportConfig = {
  api: ApiConfig
  s3: S3Config
  reportsBucket: string
  reportName: string
}

type S3Credentails = {
  accessKeyId?: string
  secretAccessKey?: string
}

const s3Credentials: S3Credentails = {}
if (process.env.S3_AWS_ACCESS_KEY_ID) {
  s3Credentials.accessKeyId = process.env.S3_AWS_ACCESS_KEY_ID
}
if (process.env.S3_AWS_ACCESS_KEY_ID) {
  s3Credentials.secretAccessKey = process.env.S3_AWS_ACCESS_KEY_ID
}

const config: AutomationReportConfig = {
  api: {
    apiUrl: process.env.API_URL ?? "http://localhost:20001",
    apiKey: process.env.API_KEY ?? "test-key"
  },
  s3: {
    endpoint: process.env.S3_ENDPOINT ?? "https://s3.eu-west-2.amazonaws.com",
    region: process.env.S3_REGION ?? "eu-west-2",
    s3ForcePathStyle: true,
    ...s3Credentials
  },
  reportsBucket: process.env.REPORTS_BUCKET ?? "bichard-7-testing-reporting-files",
  reportName: process.env.REPORT_NAME ?? "reports/AutomationRate.xlsx"
}

export default config
