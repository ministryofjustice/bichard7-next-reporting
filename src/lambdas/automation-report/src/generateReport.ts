import { findForceName } from "@bichard/forces"
import type { AuditLog, AuditLogEvent, KeyValuePair } from "@bichard/types"
import { stringify } from "csv-stringify/sync"

interface Force {
  exceptions: number
  manuallyResolved: number
  resubmittedAndResolved: number
  automated: number
  automatedQuotient: number
  automatedWithResubmissionQuotient: number
}

interface CalculateForceResult {
  force: Force
  national: Force
}

type Forces = KeyValuePair<string, Force>

const exceptionsCategoryEventTypes = [
  "Hearing Outcome passed to Error List",
  "PNC Update added to Error List",
  "Exception marked as resolved by user"
]
const manuallyResolvedCategoryEventType = "Exception marked as resolved by user"
const pncUpdateSuccessfullyEventType = "PNC Update applied successfully"

const calculateForce = (events: AuditLogEvent[], force: Force, national: Force): CalculateForceResult => {
  const sortedEvents = events.sort((eventA, eventB) => (eventA.timestamp > eventB.timestamp ? 1 : -1))
  const updatedForce = { ...force }
  const updatedNational = { ...national }

  let foundException = false
  sortedEvents.forEach((event) => {
    const { eventType } = event

    if (exceptionsCategoryEventTypes.includes(eventType)) {
      updatedForce.exceptions += 1
      updatedNational.exceptions += 1
      foundException = true
    } else if (eventType === manuallyResolvedCategoryEventType) {
      updatedForce.manuallyResolved += 1
      updatedNational.manuallyResolved += 1
    } else if (eventType === pncUpdateSuccessfullyEventType) {
      updatedForce.automated += 1
      updatedNational.automated += 1

      if (foundException) {
        updatedForce.resubmittedAndResolved += 1
        updatedNational.resubmittedAndResolved += 1
      }
    }
  })

  return { force: updatedForce, national: updatedNational }
}

const calculateForces = (messages: AuditLog[]): Forces => {
  const forces: KeyValuePair<string, Force> = {}
  const newForce = () => ({
    exceptions: 0,
    manuallyResolved: 0,
    resubmittedAndResolved: 0,
    automated: 0,
    automatedQuotient: 0,
    automatedWithResubmissionQuotient: 0
  })

  let national = newForce()

  messages.forEach((message) => {
    const { forceOwner, events } = message.automationReport
    const forceName = findForceName(forceOwner)

    if (!forceName) {
      return
    }

    const force = forces[forceName] || newForce()

    const calculateForceResult = calculateForce(events, force, national)
    national = calculateForceResult.national
    forces[forceName] = calculateForceResult.force
  })

  forces.national = national

  Object.keys(forces).forEach((forceName) => {
    const force = forces[forceName]
    const { automated, exceptions, manuallyResolved, resubmittedAndResolved } = force
    const automatedWithResubmissionDivider = automated + exceptions - manuallyResolved

    if (automatedWithResubmissionDivider > 0) {
      force.automatedQuotient = automated / automatedWithResubmissionDivider
      force.automatedWithResubmissionQuotient = (automated + resubmittedAndResolved) / automatedWithResubmissionDivider
    }
  })

  return forces
}

const generateCsv = (forces: Forces): string => {
  const lines = [["Force", "Automated", "Resubmitted", "Automated including Resubmissions"]]
  const addLine = (forceName: string, name?: string) => {
    const force = forces[forceName]
    const automated = force.automatedQuotient.toFixed(5)
    const resubmittedAndResolved = force.resubmittedAndResolved.toFixed(5)
    const automatedWithResubmission = force.automatedWithResubmissionQuotient.toFixed(5)
    lines.push([`${name || forceName}`,`${automated}`,`${resubmittedAndResolved}`,`${automatedWithResubmission}`])
  }

  Object.keys(forces)
    .filter((forceName) => forceName !== "national")
    .sort()
    .forEach((forceName: string) => addLine(forceName))

  addLine("national", "National Average")

  //return lines.join("\n")
  return stringify(lines)
}

export default (events: AuditLog[]): string => {
  const forces = calculateForces(events)

  const csvResult = generateCsv(forces)

  return csvResult
}
