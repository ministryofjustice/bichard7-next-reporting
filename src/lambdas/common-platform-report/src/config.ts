import type { DynamoDbConfig } from "@bichard/dynamo-gateway"
import { createDynamoDbConfig } from "@bichard/dynamo-gateway"
import type { SmtpConfig } from "./getEmailer"

export interface CommonPlatformReportConfig {
  smtp: SmtpConfig
  dynamo: DynamoDbConfig
  fromAddress: string
  toAddresses: string
  timePeriodHours: number
}

const config: CommonPlatformReportConfig = {
  smtp: {
    host: process.env.SMTP_HOST ?? "console",
    user: process.env.SMTP_USER ?? "bichard",
    password: process.env.SMTP_PASSWORD ?? "password",
    port: parseInt(process.env.SMTP_PORT ?? "587", 10),
    tls: process.env.SMTP_TLS === "true",
    debug: process.env.SMTP_DEBUG === "true"
  },
  dynamo: createDynamoDbConfig(),
  fromAddress: process.env.FROM_ADDRESS ?? "no-reply@mail.bichard7.service.justice.gov.uk",
  timePeriodHours: parseInt(process.env.TIME_PERIOD_HOURS ?? "240", 10),
  toAddresses: process.env.TO_ADDRESSES ?? "moj-bichard7@madetech.cjsm.net"
}

export default config
