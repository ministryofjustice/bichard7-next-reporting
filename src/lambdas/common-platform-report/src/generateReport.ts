import { stringify } from "csv-stringify/sync"
import type { ReportRecord } from "./getReportData"

export default (records: ReportRecord[]): string => {
  const report = [
    ["Received Date", "Internal Message ID", "External Correlation ID", "PTIURN", "Error Message"]
  ].concat(records.map((r) => [r.receivedDate, r.messageId, r.externalCorrelationId, r.ptiurn, r.error]))
  return stringify(report)
}
