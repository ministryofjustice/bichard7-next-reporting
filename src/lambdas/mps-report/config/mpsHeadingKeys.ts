export const headers =[
    "Hearing Date",
    "Court",
    "Force",
    "Urgent",
    "Defendant",
    "DoB",
    "ASN",
    "URN",
    "Triggers",
    "Errors",
    "Offence Code",
    "Offence Start Date (Incident date)",
    "Offence Location",
    "1st Result Code",
    "1st Result Duration / Amount",
    "1st Result Text",
    "1st Result Qualifiers",
    "2nd Result Code",
    "2nd Result Duration / Amount",
    "2nd Result Text",
    "2nd Result Qualifiers",
    "3rd Result Code",
    "3rd Result Duration / Amount",
    "3rd Result Text",
    "3rd Result Qualifiers",
    "4th Result Code",
    "4th Result Duration / Amount",
    "4th Result Text",
    "4th Result Qualifiers",
    "5th Result Code",
    "5th Result Duration / Amount",
    "5th Result Text",
    "5th Result Qualifiers",
    "6th Result Code",
    "6th Result Duration / Amount",
    "6th Result Text",
    "6th Result Qualifiers",
    "7th Result Code",
    "7th Result Duration / Amount",
    "7th Result Text",
    "7th Result Qualifiers",
    "8th Result Code",
    "8th Result Duration / Amount",
    "8th Result Text",
    "8th Result Qualifiers",
    "9th Result Code",
    "9th Result Duration / Amount",
    "9th Result Text",
    "9th Result Qualifiers",
    "10th Result Code",
    "10th Result Duration / Amount",
    "10th Result Text",
    "10th Result Qualifiers",
    "Trigger Locked By",
    "Exception Locked By"
]
