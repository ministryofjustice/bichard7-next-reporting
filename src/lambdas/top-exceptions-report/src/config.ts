import type { ApiConfig, S3Config } from "@bichard/types"

export type TopExceptionsReportConfig = {
  api: ApiConfig
  s3: S3Config
  reportsBucket: string
}

type S3Credentails = {
  accessKeyId?: string
  secretAccessKey?: string
}

const s3Credentials: S3Credentails = {}
if (process.env.S3_AWS_ACCESS_KEY_ID) {
  s3Credentials.accessKeyId = process.env.S3_AWS_ACCESS_KEY_ID
}
if (process.env.S3_AWS_ACCESS_KEY_ID) {
  s3Credentials.secretAccessKey = process.env.S3_AWS_ACCESS_KEY_ID
}

const config: TopExceptionsReportConfig = {
  api: {
    apiUrl: process.env.API_URL ?? "http://localhost:20001",
    apiKey: process.env.API_KEY ?? "test-key"
  },
  s3: {
    endpoint: process.env.S3_ENDPOINT ?? "https://s3.eu-west-2.amazonaws.com",
    region: process.env.S3_REGION ?? "eu-west-2",
    s3ForcePathStyle: true,
    ...s3Credentials
  },
  reportsBucket: process.env.REPORTS_BUCKET ?? "bichard-7-testing-reporting-files"
}

export default config
