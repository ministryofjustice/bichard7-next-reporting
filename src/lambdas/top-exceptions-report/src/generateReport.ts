import { findForceName } from "@bichard/forces"
import type { AuditLog, KeyValuePair } from "@bichard/types"
import getErrorName from "./errorNames/getErrorName"
import { stringify } from "csv-stringify/sync"

type ForceExceptions = KeyValuePair<string, number>
type ForcesExceptions = KeyValuePair<string, ForceExceptions>

const getAttributesWithErrors = (messages: AuditLog[]): KeyValuePair<string, unknown>[] => {
  return messages
    .filter((x) => x.events && x.events.some((e) =>'Error 1 Details' in e.attributes))
    .map((x) => {
      const events = x.events
      for(let i=0; i< events.length; i+=1) {
        if(x.automationReport?.forceOwner) {
          events[i].attributes["Force Owner"] = x.automationReport.forceOwner
        }
      }
      return events
    })
    .reduce((allEvents, topExceptionsReportEvents) => allEvents.concat(topExceptionsReportEvents), [])
    .map((event) => event.attributes)
}

const calculateForcesExceptions = (allAttributes: KeyValuePair<string, unknown>[]): ForcesExceptions => {
  return allAttributes.reduce((forces, attributes) => {
    const forceOwner = attributes["Force Owner"] as string
    const forceName = findForceName(forceOwner) ?? "National"
    const forceExceptions = (forces[forceName] || {}) as ForceExceptions

    Object.keys(attributes)
      .filter((attributeName) => attributeName.match(/Error.*Details/))
      .forEach((key) => {
        const exceptionCode = (attributes[key] as string).split("||")[0]
        forceExceptions[exceptionCode] = (forceExceptions[exceptionCode] || 0) + 1
      })

    // eslint-disable-next-line no-param-reassign
    forces[forceName] = forceExceptions
    return forces
  }, {}) as ForcesExceptions
}

const generateCsv = (forcesExceptions: ForcesExceptions): string => {
  let rows = [["Force" ,"Exception" ,"Error Text" ,"Count"]]
  Object.keys(forcesExceptions).forEach((forceName) => {
    const forceExceptions = forcesExceptions[forceName]
    Object.keys(forceExceptions).forEach((exceptionCode) => {
      const errorText = getErrorName(exceptionCode)
      rows.push([`${forceName}`,`${exceptionCode}`,`${errorText}`,`${forceExceptions[exceptionCode]}`])
    })
  })
  return stringify(rows)
}

export default (messages: AuditLog[]): string => {
  const attributesWithErrors = getAttributesWithErrors(messages)

  const forcesExceptions = calculateForcesExceptions(attributesWithErrors)

  const csvResult = generateCsv(forcesExceptions)

  return csvResult
}
