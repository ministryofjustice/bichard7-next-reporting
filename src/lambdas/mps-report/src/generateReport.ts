import getCourtErrors from "./getCourtErrors"
import { headers } from "../config/mpsHeadingKeys"
import getOffenceCode from "./getOffenceCode"
import AnnotatedHearingOutcome from "./types/AnnotatedHearingOutcome"
import { PostgresGateway } from "@bichard/postgres-gateway"
import { isError } from "@bichard/types"
import { xml2js } from "xml-js"
import { getText } from "./types/XmlStringCell"
import getResultQualifierCodes from "./getResultQualifierCodes"
import getDurationAndAmount from "./getDurationAndAmount"
import OffenceDetails, { isMultiple } from "./types/OffenceDetails"
import OffenceResult from "./types/OffenceResult"
import HearingOutcomeCase from "./types/HearingOutcomeCase"
import AnnotatedPNCUpdateDataset from "./types/AnnotatedPNCUpdateDataset"

interface ReportRowResultQuery {
    court_date: string
    court_code: string
    force_code: string
    is_urgent: string
    defendant_name: string
    triggers: string
    error_report: string
}

function updateCommonHeaders(row: ReportRowResultQuery, hearingOutcomeCase: HearingOutcomeCase, offence:OffenceDetails) {
  return [
    row.court_date, // Hearing Date
    row.court_code, // Court
    row.force_code, // Force
    row.is_urgent, // Urgent
    row.defendant_name, // Defendant
    getText(hearingOutcomeCase.HearingDefendant.DefendantDetail?.BirthDate), // DoB
    getText(hearingOutcomeCase.HearingDefendant.ArrestSummonsNumber), // ASN
    getText(hearingOutcomeCase.PTIURN), // URN
    row.triggers, // Triggers
    row.error_report, // Errors
    getOffenceCode(offence.CriminalProsecutionReference), // Offence Code
    getText(offence.ActualOffenceStartDate.StartDate), // Offence Start Date
    getText(offence.LocationOfOffence), // Offence Location
  ]
}

export default async (gateway: PostgresGateway) => {
  const rows = await getCourtErrors(gateway)
  if (isError(rows)) {
    return rows
  }
  
  let result = []
  result.push(headers.join(","))
  for (let i = 0; i < rows.length; i = i + 1) {
    const annotatedMsg: string = rows[i].annotated_msg.replace(/br7:/g, "").replace(/ds:/g, "")
    let annotatedMsgObject = (xml2js(annotatedMsg, { compact: true }) as AnnotatedPNCUpdateDataset)?.AnnotatedPNCUpdateDataset?.PNCUpdateDataset
    if(!annotatedMsgObject) {
      annotatedMsgObject = (xml2js(annotatedMsg, { compact: true }) as AnnotatedHearingOutcome)
    }
    const offences = annotatedMsgObject.AnnotatedHearingOutcome.HearingOutcome.Case.HearingDefendant.Offence
    const hearingOutcomeCase = annotatedMsgObject.AnnotatedHearingOutcome.HearingOutcome.Case

    if (isMultiple<OffenceDetails>(offences)) {
      for (let o = 0; o < offences.length; o = o + 1) {
        const newRow = updateCommonHeaders(rows[i], hearingOutcomeCase, offences[o])

        const results = offences[o].Result
        if (isMultiple<OffenceResult>(results)) {
          for (let j = 0; j < 10; j = j + 1) {
            if (j >= results.length) {
              newRow.push("") // result code
              newRow.push("") // result duration/amount
              newRow.push("") // result text
              newRow.push("") // result qualifier
            } else {
              // j-th
              newRow.push(getText(results[j].CJSresultCode)) // result code
              newRow.push(getDurationAndAmount(results[j])) // result duration/amount
              newRow.push(getText(results[j].ResultVariableText)) // result text
              newRow.push(getResultQualifierCodes(results[j].ResultQualifierVariable)) // result qualifier
            }
          }
        } else {
          newRow.push(getText(results.CJSresultCode)) // result code
          newRow.push(getDurationAndAmount(results)) // result duration/amount
          newRow.push(getText(results.ResultVariableText)) // result text
          newRow.push(getResultQualifierCodes(results.ResultQualifierVariable)) // result qualifier

          // fill in the remaining 9 groups of cells
          for (let j = 0; j < 9; j = j + 1) {
            newRow.push("") // result code
            newRow.push("") // result duration/amount
            newRow.push("") // result text
            newRow.push("") // result qualifier
          }
        }
        newRow.push(rows[i].trigger_locked_by_id) // Trigger Locked By
        newRow.push(rows[i].error_locked_by_id) // Exception Locked By

        result.push(newRow.join(","))
      }
    } else {
      const newRow = updateCommonHeaders(rows[i], hearingOutcomeCase, offences)

      const results = offences.Result
      if (isMultiple<OffenceResult>(results)) {
        for (let j = 0; j < 10; j = j + 1) {
          if (j >= results.length) {
            newRow.push("") // result code
            newRow.push("") // result duration/amount
            newRow.push("") // result text
            newRow.push("") // result qualifier
          } else {
            // j-th
            newRow.push(getText(results[j].CJSresultCode)) // result code
            newRow.push(getDurationAndAmount(results[j])) // result duration/amount
            newRow.push(getText(results[j].ResultVariableText)) // result text
            newRow.push(getResultQualifierCodes(results[j].ResultQualifierVariable)) // result qualifier
          }
        }
      } else {
        newRow.push(getText(results.CJSresultCode)) // result code
        newRow.push(getDurationAndAmount(results)) // result duration/amount
        newRow.push(getText(results.ResultVariableText)) // result text
        newRow.push(getResultQualifierCodes(results.ResultQualifierVariable)) // result qualifier

        // fill in the remaining 9 groups of cells
        for (let j = 0; j < 9; j = j + 1) {
          newRow.push("") // result code
          newRow.push("") // result duration/amount
          newRow.push("") // result text
          newRow.push("") // result qualifier
        }
      }
      newRow.push(rows[i].trigger_locked_by_id) // Trigger Locked By
      newRow.push(rows[i].error_locked_by_id) // Exception Locked By

      result.push(newRow.join(","))
    }
  }

  return result.join("\r\n")
}
