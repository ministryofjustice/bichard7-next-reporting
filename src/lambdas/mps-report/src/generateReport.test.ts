import { PostgresGateway } from "@bichard/postgres-gateway"
import generateReport from "./generateReport"
import config from "./lib/config"
import {readFileSync} from 'fs'

describe("GenerateReport", () => {
    let gateway: PostgresGateway

    beforeAll(async () => {
        gateway = new PostgresGateway(config.database)
    })

    beforeEach(async ()=> {
        await gateway.getResult(`DELETE FROM br7own.error_list`)
    })

    it("should generate report when errors are present in the table 1", async () => {
        let errorReport = readFileSync("src/exampleFiles/exampleAnnotatedMsg.xml", 'utf8')

        errorReport = errorReport.replace(/'/g,"")
        await gateway.getResult(`INSERT INTO br7own.error_list 
        VALUES(default, 'msgid1', 0, 1, 1, 0, 0, 0, 'System', null, 42, 'ASN', 'FFUUSS', '${errorReport}', 'UPD-MSG', '', current_timestamp, 'error-reason', 'trg-reason', 0, 0, current_date, 'ptiurn', 'court-name', null, current_timestamp, 'err-resolved-by', 'trg-resolved-by', null, null, 'defendant', 'FFUUSS', null, 'court-ref', null, null, null, default, default)`)
        await gateway.getResult(`
            UPDATE br7own.error_list
            SET court_date = '9/1/2019'
            WHERE message_id = 'msgid1'
       `)
        const reportResult = await generateReport(gateway)
        expect(reportResult).not.toBe(undefined)
            expect(reportResult).toBe(`Hearing Date,Court,Force,Urgent,Defendant,DoB,ASN,URN,Triggers,Errors,Offence Code,Offence Start Date (Incident date),Offence Location,1st Result Code,1st Result Duration / Amount,1st Result Text,1st Result Qualifiers,2nd Result Code,2nd Result Duration / Amount,2nd Result Text,2nd Result Qualifiers,3rd Result Code,3rd Result Duration / Amount,3rd Result Text,3rd Result Qualifiers,4th Result Code,4th Result Duration / Amount,4th Result Text,4th Result Qualifiers,5th Result Code,5th Result Duration / Amount,5th Result Text,5th Result Qualifiers,6th Result Code,6th Result Duration / Amount,6th Result Text,6th Result Qualifiers,7th Result Code,7th Result Duration / Amount,7th Result Text,7th Result Qualifiers,8th Result Code,8th Result Duration / Amount,8th Result Text,8th Result Qualifiers,9th Result Code,9th Result Duration / Amount,9th Result Text,9th Result Qualifiers,10th Result Code,10th Result Duration / Amount,10th Result Text,10th Result Qualifiers,Trigger Locked By,Exception Locked By
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,1948-11-11,1101ZD0100000448754K,01ZD0303208,,,SX03001A,2010-11-28,Kingston High Street,3078,,Travel Restriction Order,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,1948-11-11,1101ZD0100000448754K,01ZD0303208,,,SX03001,2010-11-28,Kingston High Street,3052,,defendant must never be allowed out,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,1948-11-11,1101ZD0100000448754K,01ZD0303208,,,RT88191,2010-11-28,Kingston High Street,1015,100.00,Fined 100.,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System`)
    })

    it("should generate report when using record with multiple durations", async () => {
        let errorReport = readFileSync("src/exampleFiles/exampleMultipleDurationMsg.xml", 'utf8')

        errorReport = errorReport.replace(/'/g,"")
        await gateway.getResult(`INSERT INTO br7own.error_list 
        VALUES(default, 'msgid1', 0, 1, 1, 0, 0, 0, 'System', null, 42, 'ASN', 'FFUUSS', '${errorReport}', 'UPD-MSG', '', current_timestamp, 'error-reason', 'trg-reason', 0, 0, current_date, 'ptiurn', 'court-name', null, current_timestamp, 'err-resolved-by', 'trg-resolved-by', null, null, 'defendant', 'FFUUSS', null, 'court-ref', null, null, null, default, default)`)
        await gateway.getResult(`
            UPDATE br7own.error_list
            SET court_date = '9/1/2019'
            WHERE message_id = 'msgid1'
       `)
        const reportResult = await generateReport(gateway)
        expect(reportResult).not.toBe(undefined)
            expect(reportResult).toBe(`Hearing Date,Court,Force,Urgent,Defendant,DoB,ASN,URN,Triggers,Errors,Offence Code,Offence Start Date (Incident date),Offence Location,1st Result Code,1st Result Duration / Amount,1st Result Text,1st Result Qualifiers,2nd Result Code,2nd Result Duration / Amount,2nd Result Text,2nd Result Qualifiers,3rd Result Code,3rd Result Duration / Amount,3rd Result Text,3rd Result Qualifiers,4th Result Code,4th Result Duration / Amount,4th Result Text,4th Result Qualifiers,5th Result Code,5th Result Duration / Amount,5th Result Text,5th Result Qualifiers,6th Result Code,6th Result Duration / Amount,6th Result Text,6th Result Qualifiers,7th Result Code,7th Result Duration / Amount,7th Result Text,7th Result Qualifiers,8th Result Code,8th Result Duration / Amount,8th Result Text,8th Result Qualifiers,9th Result Code,9th Result Duration / Amount,9th Result Text,9th Result Qualifiers,10th Result Code,10th Result Duration / Amount,10th Result Text,10th Result Qualifiers,Trigger Locked By,Exception Locked By
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,,0000000000000000000A,00000000000,,,AA00000A,1967-08-13,a,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,,0000000000000000000A,00000000000,,,AA00000A,1967-08-13,a,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,,0000000000000000000A,00000000000,,,AA00000A,1967-08-13,a,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System`)
    })
})
