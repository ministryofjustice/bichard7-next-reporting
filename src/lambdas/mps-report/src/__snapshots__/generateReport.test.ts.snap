// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerateReport should generate report when errors are present in the table 1 1`] = `
"Hearing Date,Court,Force,Urgent,Defendant,DoB,ASN,URN,Triggers,Errors,Offence Code,Offence Start Date (Incident date),Offence Location,1st Result Code,1st Result Duration / Amount,1st Result Text,1st Result Qualifiers,2nd Result Code,2nd Result Duration / Amount,2nd Result Text,2nd Result Qualifiers,3rd Result Code,3rd Result Duration / Amount,3rd Result Text,3rd Result Qualifiers,4th Result Code,4th Result Duration / Amount,4th Result Text,4th Result Qualifiers,5th Result Code,5th Result Duration / Amount,5th Result Text,5th Result Qualifiers,6th Result Code,6th Result Duration / Amount,6th Result Text,6th Result Qualifiers,7th Result Code,7th Result Duration / Amount,7th Result Text,7th Result Qualifiers,8th Result Code,8th Result Duration / Amount,8th Result Text,8th Result Qualifiers,9th Result Code,9th Result Duration / Amount,9th Result Text,9th Result Qualifiers,10th Result Code,10th Result Duration / Amount,10th Result Text,10th Result Qualifiers,Trigger Locked By,Exception Locked By
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,1948-11-11,1101ZD0100000448754K,01ZD0303208,,,SX03001A,2010-11-28,Kingston High Street,3078,,Travel Restriction Order,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,1948-11-11,1101ZD0100000448754K,01ZD0303208,,,SX03001,2010-11-28,Kingston High Street,3052,,defendant must never be allowed out,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,1948-11-11,1101ZD0100000448754K,01ZD0303208,,,RT88191,2010-11-28,Kingston High Street,1015,100.00,Fined 100.,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System"
`;

exports[`GenerateReport should generate report when using record with multiple durations 1`] = `
"Hearing Date,Court,Force,Urgent,Defendant,DoB,ASN,URN,Triggers,Errors,Offence Code,Offence Start Date (Incident date),Offence Location,1st Result Code,1st Result Duration / Amount,1st Result Text,1st Result Qualifiers,2nd Result Code,2nd Result Duration / Amount,2nd Result Text,2nd Result Qualifiers,3rd Result Code,3rd Result Duration / Amount,3rd Result Text,3rd Result Qualifiers,4th Result Code,4th Result Duration / Amount,4th Result Text,4th Result Qualifiers,5th Result Code,5th Result Duration / Amount,5th Result Text,5th Result Qualifiers,6th Result Code,6th Result Duration / Amount,6th Result Text,6th Result Qualifiers,7th Result Code,7th Result Duration / Amount,7th Result Text,7th Result Qualifiers,8th Result Code,8th Result Duration / Amount,8th Result Text,8th Result Qualifiers,9th Result Code,9th Result Duration / Amount,9th Result Text,9th Result Qualifiers,10th Result Code,10th Result Duration / Amount,10th Result Text,10th Result Qualifiers,Trigger Locked By,Exception Locked By
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,,0000000000000000000A,00000000000,,,AA00000A,1967-08-13,a,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,,0000000000000000000A,00000000000,,,AA00000A,1967-08-13,a,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System
Sun Sep 01 2019 00:00:00 GMT+0100 (British Summer Time),FFUUSS,FFUUSS,42,defendant,,0000000000000000000A,00000000000,,,AA00000A,1967-08-13,a,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,1001,1H 1H 1H 0.01 0.01 0.01,a,  ,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,System"
`;
